clang/lib/Sema/SemaDecl.cpp

```diff
@@ -20134,9 +20134,10 @@ Decl *Sema::ActOnEnumConstant(Scope *S, Decl *theEnumDecl, Decl *lastEnumConst,
   // different from T:
   // - every enumerator of every member of class T that is an unscoped
   // enumerated type
-  if (getLangOpts().CPlusPlus && !TheEnumDecl->isScoped())
-    DiagnoseClassNameShadow(TheEnumDecl->getDeclContext(),
-                            DeclarationNameInfo(Id, IdLoc));
+  if (getLangOpts().CPlusPlus && !TheEnumDecl->isScoped() &&
+      DiagnoseClassNameShadow(TheEnumDecl->getDeclContext(),
+                              DeclarationNameInfo(Id, IdLoc)))
+    return nullptr;
```

Comment:
Perhaps a test case could be added along the lines of:
```
struct S {
  static constexpr enum E {
    S = 12
  } e = S;
  static_assert(e == S);
  static_assert(S == 12);
};
```
I believe the existing code passes because we diagnose the shadow but then go ahead and make an enumerator with the class name, and oofda is that confusing: https://godbolt.org/z/K4c83frMh